File: leetcodePOTD/04-march-2024.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        n=len(tokens)
        left,right=0,n-1
        maxscore=0
        score=0
        while left <=right:
            if power>=tokens[left]:
                power-=tokens[left]
                score+=1
                maxscore=max(maxscore,score)
                left+=1
            elif score>0:
                power+=tokens[right]
                score-=1
                right-=1
            else:
                break
        return maxscore                

File: leetcodePOTD/04-march-2024.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        int n =tokens.size();
        int i=0,j=n-1;
        sort(begin(tokens),end(tokens));
        int score=0;
        int maxs=0;
        while(i<=j){
            if(power>=tokens[i]){
                power-=tokens[i];
                score+=1;
                i++;
                maxs=max(maxs,score);
            }
            else if(score>=1){
                power+=tokens[j];
                j--;
                score-=1;
            }
            else{
                return maxs;
            }
        }return maxs;
        }
        

        
    
};